### ISSUE TITLE
Add context for query executions

### ISSUE THREAD

**Issue by salsakran** [metabase/metabase#4522]
As part of being able to control embedding, we should add context to the query execution log as to whether a query executions was:

0) due to a database sync or other system operation
1) in a "normal" dashboard view
2) in a public dashboard
3) in an embedded dashboard 
4) in a "normal" question view
5) in a public card view
6) in an embedded dashboard
7) as a CSV download
8) as a JSON download
9) as a pulse rendering

most efficient thing is probably to use use an int/enum for this vs putting it in the context dictionary.

**Comment by camsaul** [comment]
Adding a new column to the query execution table could potentially take a while when we have millions of rows. Are we ok with that?

(vs just using the `:additional_info` dictionary to store it)

**Comment by salsakran** [comment]
I think to un-fuck the queryexecution table situation we'll need to modify this table.

Depending on the extent to changes, a new table with the right schema + a partial copy might be the best way forward as you mentioned. However, I'd suggest just going ahead with this PR, and we can try it on a couple of instance application dbs we have lying around to see how bad it is in real life.

**Comment by camsaul** [comment]
yeah, I actually have an even more aggressive version of this PR on my computer that I was playing around with this morning. There's a lot of other changes that I think make sense:

*  There's an index on the `executor_id` column that is basically useless. It would be useful if we were doing something like looking up the queries ran by a certain user, but we're not; it made sense a while back when we actually deleted users (meaning we had to cascade-delete the query executions due to the FK constraint) but that's not the case anymore. Indexes take up a lot of space when you have huge tables (not to mention slowing down `INSERT`s, which makes a big difference here since we `INSERT` into this table *all the time*) so it's something I'd drop if we did it over. 
*  The `json_query` column isn't useful and wastes a lot of space, we should just stick to using query hashes
*  We *should* have an index on `query_hash + started_at`, to facilitate calculating average historic execution times (e.g., for the caching PR). Right now we have individual indexes on both columns. This isn't quite as good: for example, to find the 10 most recent executions of a given query, what happens is that the DB can use one of those indexes, probably `query_hash`, to narrow down the set of records, and then do a scan over that subset of records to find the 10 most recent. If we had a 2-column index we could just head straight to the relevant hash and pick of 10 records, already in order. FWIW I made this same exact point way back in the day :/
*  Using the "secure hash" from the caching PR everywhere would be nice, as opposed to using the "simple hash" in some places (QueryExecutions) and a different hash for caching. As @tlrobinson pointed out, [Java's `hashCode()` is not guaranteed to be the same across multiple JVM launches](http://eclipsesource.com/blogs/2012/09/04/the-3-things-you-should-know-about-hashcode/). From what I understand Clojure uses a combination of Murmur and `hashCode()` for hashing, depending on the object type; I haven't seen a case where the query hashes change between launches but I can guarantee that they dont :( 

I think the best option would be to create a new table, and possibly `SELECT INTO` that table with some subset of records (if we kept the old hash style). We would probably just have to start a new era of history if we wanted to switch over to using the "secure hash", which I think would be preferable in many ways going forward, but would unfortunately be like the QueryExecution equivalent of the end of Fight Club where everyone's credit history is wiped clean

**Comment by camsaul** [comment]
Either way a new table would be preferable since it seems like you can't drop an index that "belongs to" a foreign key in H2. i.e. I can do 

```sql
ALTER TABLE query_queryexecution
DROP INDEX idx_queryexecution_executor_id
```

on Postgres or MySQL without issues but H2 doesn't like it

**Comment by camsaul** [comment]
Implemented by #4351
